src/Particular.HealthMonitoring.Uptime/EventUnitOfWork.cs

```diff
@@ -0,0 +1,38 @@
+namespace Particular.HealthMonitoring.Uptime
+{
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Particular.HealthMonitoring.Uptime.Api;
+    using ServiceControl.Infrastructure.DomainEvents;
+
+    class EventUnitOfWork : IDomainEvents
+    {
+        IDomainEvents domainEvents;
+        IPersistEndpointUptimeInformation persister;
+        List<IHeartbeatEvent> recordedEvents = new List<IHeartbeatEvent>();
+
+        public EventUnitOfWork(IDomainEvents domainEvents, IPersistEndpointUptimeInformation persister)
+        {
+            this.domainEvents = domainEvents;
+            this.persister = persister;
+        }
+
+        public void Raise<T>(T domainEvent) where T : IDomainEvent
+        {
+            domainEvents.Raise(domainEvent);
+            var heartbeatEvent = domainEvent as IHeartbeatEvent;
+            if (heartbeatEvent != null)
+            {
+                recordedEvents.Add(heartbeatEvent);
+            }
+        }
+
+        public async Task Persist()
+        {
+            foreach (var @event in recordedEvents)
+            {
+                await persister.Store(@event).ConfigureAwait(false);
```

Comment:
I'm struggling to follow this through, but how many of these would be expected at a time? For the most part I see these as single operations except for the timer that would invoke this every 5 seconds. Given that this could be called every 5 seconds for every endpoint, is it worth creating a persistence Unit of Work so that the `SaveChangesAsync` is only called once and the operation is batched?